README
NAME

Lexical::SealRequireHints - prevent leakage of lexical hints

DESCRIPTION

This module works around two historical bugs in Perl's handling of the
"%^H" (lexical hints) variable.  One bug causes lexical state in one file
to leak into another that is "require"d/"use"d from it.  This bug, [perl
#68590], was present up to Perl 5.10, fixed in Perl 5.11.0.  The second
bug causes lexical state (normally a blank "%^H" once the first bug is
fixed) to leak outwards from "utf8.pm", if it is automatically loaded
during Unicode regular expression matching, into whatever source is
compiling at the time of the regexp match.  This bug, [perl #73174],
was present from Perl 5.8.7 up to Perl 5.11.5, fixed in Perl 5.12.0.

Both of these bugs seriously damage the usability of any module
relying on "%^H" for lexical scoping, on the affected Perl versions.
It is in practice essential for such modules to work around these bugs.
On versions of Perl that require such a workaround, this module globally
changes the behaviour of "require", including "use" and the implicit
"require" performed in Unicode regular expression matching, so that
it no longer exhibits these bugs.  This is the most convenient kind of
workaround, and is meant to be invoked by any module that makes use of
lexical state.

The workaround supplied by this module takes effect the first time its
"import" method is called.  Typically this will be done by means of
a "use" statement.  This should be done before putting anything into
"%^H" that would have a problem with leakage; usually it suffices to do
this when loading the module that supplies the mechanism to set up the
vulnerable lexical state.  Invoking this module multiple times, from
multiple lexical-related modules, is not a problem: the workaround is
only applied once, and applies to everything.

This module is implemented in XS, with a pure Perl backup version for
systems that can't handle XS modules.  The XS version has a better chance
of playing nicely with other modules that modify "require" handling.

INSTALLATION

	perl Build.PL
	./Build
	./Build test
	./Build install

AUTHOR

Andrew Main (Zefram) <zefram@fysh.org>

COPYRIGHT

Copyright (C) 2009, 2010, 2011 Andrew Main (Zefram) <zefram@fysh.org>

LICENSE

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
version 0.006; 2011-11-20

  * bugfix: avoid loading warnings.pm and leaving its delayed requires
    of Carp.pm susceptible to hint leakage, which was causing trouble
    on some Perls

  * skip swash test on Perl 5.6, where swash loading appears to be broken
    by loading Test::More or anything else useful

  * remove bogus tests that cause false failures on Perl 5.15.5

  * in Build.PL, declare incompatibility with pre-0.19
    B::Hooks::OP::Check, which doesn't play nicely around op check hooking

  * comment why a variable surprisingly needs to be "our"

  * convert .cvsignore to .gitignore

version 0.005; 2011-07-25

  * bugfix: work around core bug [perl #73174] affecting Unicode swash
    loading, and apply entire workaround arrangement to 5.11.{0..5}
    where [perl #73174] exists but [perl #68590] does not

  * correct dynamic_config setting to 0

  * include META.json in distribution

  * add MYMETA.json to .cvsignore

version 0.004; 2010-11-21

  * bugfix: don't attempt to use XS version of the workaround on Win32,
    where it can't work properly due to linker restriction on access to
    core symbols

  * only define PERL_CORE for compilation on Perl versions where the
    bug workaround (and thus interference with core-private stuff)
    is actually necessary

  * in XS, use PERL_NO_GET_CONTEXT for efficiency

  * in XS, declare "PROTOTYPES: DISABLE" to prevent automatic generation
    of unintended prototypes

  * in XS, provide a reserve definition of croak, so that the Perl_croak
    circumlocution is avoided even with PERL_CORE defined

  * in XS, give symbolic names to the Perl version thresholds

  * jump through hoops to avoid compiler warnings

  * use full stricture in test suite

  * also test POD coverage of pure Perl implementation

  * in t/setup_pp.pl, avoid a warning that occurs if XSLoader::load()
    is given no arguments, which is now a valid usage

  * in Build.PL, explicitly set needs_compiler to avoid bogus
    auto-dependency on ExtUtils::CBuilder

  * in Build.PL, complete declaration of configure-time requirements

version 0.003; 2010-04-10

  * bugfix: in pure-Perl implementation, make sure ambient package (from
    which require is invoked) is passed on correctly to the code in the
    required file, on those Perls where it is so inherited

  * in XS, use macros to avoid explicit passing of aTHX, in the manner
    of the core

  * in XS, avoid using "class" as a variable name, for compatibility
    with C++ compilers

  * make all numeric comparisons against $] stringify it first, to avoid
    architecture-dependent problems with floating point rounding giving
    it an unexpected numeric value

  * in Build.PL, explicitly declare configure-time requirements

  * add MYMETA.yml to .cvsignore

version 0.002; 2009-10-21

  * generate a more normal-looking op tree, that doesn't crash B::Deparse

  * don't apply the workaround on Perl 5.11.0 or later, where the bug
    has been fixed

  * in t/seal.t, test that cop_hints_hash is properly handled

  * check for required Perl version at runtime

version 0.001; 2009-09-26

  * bugfix: die cleanly if the pure-Perl implementation is needed but
    won't work (which occurs on pre-5.8 perls)

  * bugfix: avoid undesired warning from pure-Perl implementation if
    require has already been overridden via CORE::GLOBAL::require

  * in tests, set HINT_LOCALIZE_HH where appropriate, to avoid false
    test failures on pre-5.10 perls

  * test that the module plays nicely with code that overrides require
    via CORE::GLOBAL::require

version 0.000; 2009-09-22

  * initial released version


t/seal_0.pm
package t::seal_0;

use warnings;
use strict;

use Test::More;

BEGIN { is $^H{"Lexical::SealRequireHints/test"}, undef; }
main::test_runtime_hint_hash "Lexical::SealRequireHints/test", undef;

is $^H{"Lexical::SealRequireHints/test"}, undef;

sub import {
	is $^H{"Lexical::SealRequireHints/test"}, 1;
	$^H |= 0x20000 if "$]" < 5.009004;
	$^H{"Lexical::SealRequireHints/test0"}++;
}

1;


t/seal_1.pm
package t::seal_1;

use warnings;
use strict;

use Test::More;

BEGIN { is $^H{"Lexical::SealRequireHints/test"}, undef; }
main::test_runtime_hint_hash "Lexical::SealRequireHints/test", undef;

is $^H{"Lexical::SealRequireHints/test"}, undef;

sub import {
	is $^H{"Lexical::SealRequireHints/test"}, 1;
	$^H |= 0x20000 if "$]" < 5.009004;
	$^H{"Lexical::SealRequireHints/test1"}++;
}

1;


t/setup_pp.pl
if("$]" < 5.008) {
	require Test::More;
	Test::More::plan(skip_all =>
		"pure Perl Lexical::SealRequireHints can't work on this perl");
}

require XSLoader;

my $orig_load = \&XSLoader::load;
# Suppress redefinition warning, without loading warnings.pm, for the
# benefit of before_warnings.t.
BEGIN { ${^WARNING_BITS} = ""; }
*XSLoader::load = sub {
	die "XS loading disabled for Lexical::SealRequireHints"
		if ($_[0] || "") eq "Lexical::SealRequireHints";
	goto &$orig_load;
};

1;


Build.PL
{ use 5.006; }
use warnings;
use strict;

use Module::Build;

Module::Build->subclass(code => q{
	unless(__PACKAGE__->can("cbuilder")) {
		*cbuilder = sub { $_[0]->_cbuilder or die "no C support" };
	}
	unless(__PACKAGE__->can("have_c_compiler")) {
		*have_c_compiler = sub {
			my $cb = eval { $_[0]->cbuilder };
			return $cb && $cb->have_compiler;
		};
	}
	if($Module::Build::VERSION < 0.33) {
		# Older versions of Module::Build have a bug where if the
		# cbuilder object is used at Build.PL time (which it will
		# be for this distribution due to the logic in
		# ->find_xs_files) then that object can be dumped to the
		# build_params file, and then at Build time it will
		# attempt to use the dumped blessed object without loading
		# the ExtUtils::CBuilder class that is needed to make it
		# work.
		*write_config = sub {
			delete $_[0]->{properties}->{_cbuilder};
			return $_[0]->SUPER::write_config;
		};
	}
	sub find_xs_files {
		my($self) = @_;
		return {} unless $self->have_c_compiler;
		# On MSWin32, the XS version of the workaround can't work
		# properly, because it doesn't have access to the core
		# symbols to let SAVEHINTS() work.
		return {} if "$]" < 5.012 && $^O eq "MSWin32";
		return $self->SUPER::find_xs_files;
	}
	sub compile_c {
		my($self, $file, %args) = @_;
		if("$]" < 5.012) {
			# need PERL_CORE for working SAVEHINTS()
			$args{defines} = { %{$args{defines} || {}},
				PERL_CORE => 1,
			};
		}
		return $self->SUPER::compile_c($file, %args);
	}
})->new(
	module_name => "Lexical::SealRequireHints",
	license => "perl",
	configure_requires => {
		"Module::Build" => 0,
		"perl" => "5.006001",
		"strict" => 0,
		"warnings" => 0,
	},
	configure_recommends => {
		"ExtUtils::CBuilder" => "0.15",
	},
	build_requires => {
		"Module::Build" => 0,
		"Test::More" => 0,
		"perl" => "5.006001",
		"strict" => 0,
		"warnings" => 0,
	},
	build_recommends => {
		"ExtUtils::CBuilder" => "0.15",
	},
	requires => {
		"perl" => "5.006001",
	},
	recommends => {
		"XSLoader" => 0,
	},
	conflicts => {
		"B::Hooks::OP::Check" => "< 0.19",
	},
	needs_compiler => 0,
	dynamic_config => 0,
	meta_add => { distribution_type => "module" },
	create_makefile_pl => "passthrough",
	sign => 1,
)->create_build_script;

1;


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.3800
require 5.006001;

    unless (eval "use Module::Build::Compat 0.02; 1" ) {
      print "This module requires Module::Build to install itself.\n";

      require ExtUtils::MakeMaker;
      my $yn = ExtUtils::MakeMaker::prompt
	('  Install Module::Build now from CPAN?', 'y');

      unless ($yn =~ /^y/i) {
	die " *** Cannot install without Module::Build.  Exiting ...\n";
      }

      require Cwd;
      require File::Spec;
      require CPAN;

      # Save this 'cause CPAN will chdir all over the place.
      my $cwd = Cwd::cwd();

      CPAN::Shell->install('Module::Build::Compat');
      CPAN::Shell->expand("Module", "Module::Build::Compat")->uptodate
	or die "Couldn't install Module::Build, giving up.\n";

      chdir $cwd or die "Cannot chdir() back to $cwd: $!";
    }
    eval "use Module::Build::Compat 0.02; 1" or die $@;
    use lib '_build/lib';
    Module::Build::Compat->run_build_pl(args => \@ARGV);
    my $build_script = 'Build';
    $build_script .= '.com' if $^O eq 'VMS';
    exit(0) unless(-e $build_script); # cpantesters convention
    require MyModuleBuilder;
    Module::Build::Compat->write_makefile(build_class => 'MyModuleBuilder');


lib/Lexical/SealRequireHints.pm
=head1 NAME

Lexical::SealRequireHints - prevent leakage of lexical hints

=head1 SYNOPSIS

	use Lexical::SealRequireHints;

=head1 DESCRIPTION

This module works around two historical bugs in Perl's handling of the
C<%^H> (lexical hints) variable.  One bug causes lexical state in one file
to leak into another that is C<require>d/C<use>d from it.  This bug, [perl
#68590], was present up to Perl 5.10, fixed in Perl 5.11.0.  The second
bug causes lexical state (normally a blank C<%^H> once the first bug is
fixed) to leak outwards from C<utf8.pm>, if it is automatically loaded
during Unicode regular expression matching, into whatever source is
compiling at the time of the regexp match.  This bug, [perl #73174],
was present from Perl 5.8.7 up to Perl 5.11.5, fixed in Perl 5.12.0.

Both of these bugs seriously damage the usability of any module
relying on C<%^H> for lexical scoping, on the affected Perl versions.
It is in practice essential for such modules to work around these bugs.
On versions of Perl that require such a workaround, this module globally
changes the behaviour of C<require>, including C<use> and the implicit
C<require> performed in Unicode regular expression matching, so that
it no longer exhibits these bugs.  This is the most convenient kind of
workaround, and is meant to be invoked by any module that makes use of
lexical state.

The workaround supplied by this module takes effect the first time its
C<import> method is called.  Typically this will be done by means of
a C<use> statement.  This should be done before putting anything into
C<%^H> that would have a problem with leakage; usually it suffices to
do this when loading the module that supplies the mechanism to set up
the vulnerable lexical state.  Invoking this module multiple times,
from multiple lexical-related modules, is not a problem: the workaround
is only applied once, and applies to everything.

This module is implemented in XS, with a pure Perl backup version for
systems that can't handle XS modules.  The XS version has a better chance
of playing nicely with other modules that modify C<require> handling.

=cut

package Lexical::SealRequireHints;

{ use 5.006001; }
# Don't "use warnings" here because warnings.pm can include require
# statements that execute at runtime, and if they're compiled before
# this module takes effect then they won't get the magic needed to avoid
# leaking hints generated later.

our $VERSION = "0.006";

if("$]" >= 5.012) {
	# bug not present
	*import = sub {
		die "$_[0] does not take any importation arguments\n"
			unless @_ == 1;
	};
	*unimport = sub { die "$_[0] does not support unimportation\n" };
} elsif(eval { local $SIG{__DIE__};
	require XSLoader;
	XSLoader::load(__PACKAGE__, $VERSION);
	1;
}) {
	# Successfully loaded XS.  Now preemptively load modules that
	# may be subject to delayed require statements in XSLoader or
	# things that it loaded.
	foreach(qw(Carp.pm Carp/Heavy.pm)) {
		eval { local $SIG{__DIE__}; require($_); };
	}
} elsif("$]" >= 5.008) {
	my $done;
	*import = sub {
		die "$_[0] does not take any importation arguments\n"
			unless @_ == 1;
		return if $done;
		$done = 1;
		# $next_require empirically doesn't work as a my variable.
		our $next_require = defined(&CORE::GLOBAL::require) ?
			\&CORE::GLOBAL::require : sub {
				my($arg) = @_;
				# The shenanigans with $CORE::GLOBAL::{require}
				# are required because if there's a
				# &CORE::GLOBAL::require when the eval is
				# executed then the CORE::require in there is
				# interpreted as plain require on some Perl
				# versions, leading to recursion.
				my $grequire = $CORE::GLOBAL::{require};
				delete $CORE::GLOBAL::{require};
				my $result = eval q{
					local $SIG{__DIE__};
					$CORE::GLOBAL::{require} = $grequire;
					package }.caller(0).q{;
					CORE::require($arg);
				};
				die $@ if $@ ne "";
				return $result;
			};
		# Need to suppress the redefinition warning, without
		# invoking warnings.pm.
		BEGIN { ${^WARNING_BITS} = ""; }
		*CORE::GLOBAL::require = sub {
			die "wrong number of arguments to require\n"
				unless @_ == 1;
			my($arg) = @_;
			# %^H gets localised (in the magic way it
			# requires) by the string eval, provided that the
			# HINT_LOCALIZE_HH bit is set.	Normally that
			# bit would be set if there were anything in
			# %^H, but when affected by [perl #73174] the
			# core's swash-loading code locally clears $^H
			# without changing %^H, so we set the bit here.
			# We localise $^H while doing this, in order to
			# not clobber $^H across a normal require where
			# the bit is legitimately clear, except on Perl
			# 5.11, where the bit needs to stay set in order
			# to get proper restoration of %^H.
			local $^H = $^H | 0x20000 if "$]" < 5.011;
			$^H |= 0x20000;
			my $result = eval q{
				local $SIG{__DIE__};
				package }.caller(0).q{;
				delete $^H{$_}
					foreach keys(%^H), qw($[ open< open>);
				$next_require->($arg);
			};
			die $@ if $@ ne "";
			return $result;
		};
	};
	*unimport = sub { die "$_[0] does not support unimportation\n" };
} else {
	die "pure Perl version of @{[__PACKAGE__]} can't work on pre-5.8 perl";
}

=head1 BUGS

The operation of this module depends on influencing the compilation of
C<require>.  As a result, it cannot prevent lexical state leakage through
a C<require> statement that was compiled before this module was invoked.
This is not a problem when lexical state is managed in the usual ways:
the leakage that is a problem is almost always through C<use> statements,
which are executed immediately after they are compiled.  The situations
that would escape the sealant of this module are rather convoluted.
If such a problem does occur, a workaround is to invoke this module
earlier.

=head1 SEE ALSO

L<perlpragma>

=head1 AUTHOR

Andrew Main (Zefram) <zefram@fysh.org>

=head1 COPYRIGHT

Copyright (C) 2009, 2010, 2011 Andrew Main (Zefram) <zefram@fysh.org>

=head1 LICENSE

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;
